### PR TITLE
Trigger time based escalations

### LINKED ISSUE #63: Trigger time-based escalations without external events
At the moment, time-based escalations (something like `incident_age>=1h`) will only trigger if a new event is processed (which is not guaranteed to happen). Therefore, we have to trigger some escalations internally.

The simple fix for this would be to check all not yet escalations against open incidents periodically, but that's inefficient and introduces delays.

The nicer, but also more complex to implement solution for this would be to extract information about relevant times where the result of evaluating the escalation condition might change. Take the escalation condition `(incident_severity>=critical&incident_age>=1h)|incident_age>=3h`. Extracting each value `incident_age` is compared against gives `1h` and `3h`, therefore, the only times when it's interesting to reevaluate this condition would be `incident.started_at + 1h` and `incident.started_at + 3h` (and if the severity changes, but that is already checked when processing that event that changes this severity).

While I expect `>=` to be the most useful operator for comparing `incident_age`, we also have to consider what happens when using `=` or `>`. If we evaluate `incident_age=1h` at exactly `incident.started_at + 1h`, this will trigger (which we can't reliably achieve with periodic evaluation). Similarly, if we evaluate `incident_age>1h` at that very same time, it would not yet trigger, even though it would a tiny moment later, so we might want to evaluate it bit before and after as well (a bit = 1ns should do the trick).

### PR DESCRIPTION
**Pull request by yhabteab**
fixes #63

### CODE REVIEW

File: internal/rule/condition.go [hunk 1]
```diff
@@ -11,74 +12,92 @@ type EscalationFilter struct {
...
+			}
+
+			if incidentAge < age && (retryAfter <= 0 || retryAfter > age) {
+				retryAfter = age
+			}
+		}
+	}
+
+	return retryAfter
+}
```
Comment (julianbrost) [inline internal/rule/condition.go:31]: This function is a really good candidate for a doc comment explaining what it does.

**Review by julianbrost — COMMENTED** [review]
First few comments only, i.e. those until I noticed that I should probably look at #67 first.

File: internal/incident/incident.go [hunk 2]
```diff
@@ -595,6 +643,117 @@ func (i *Incident) processAcknowledgementEvent(ctx context.Context, tx *sqlx.Tx,
...
+
+	if err := i.ReloadRecipients(ctx); err != nil {
+		// The error is already logged within `i.ReloadRecipients()`
+		return
+	}
+
+	if !i.RecoveredAt.IsZero() {
+		// Incident is recovered in the meantime.
+		return
+	}
```
Comment (julianbrost) [inline internal/incident/incident.go:675]: This check should also be fine before performing any database query, the incident is locked so it shouldn't change.

File: internal/incident/incident.go [hunk 3]
```diff
@@ -354,20 +368,50 @@ func (i *Incident) evaluateEscalations() {
...
 						matched = false
+					} else if !matched {
+						age := cond.ExtractRetryAfter(escalation.Condition.ExtractConditions(), cond.IncidentAge)
+						if age > 0 && (retryAfter <= 0 || retryAfter > age) {
+							retryAfter = time.Until(i.StartedAt.Add(age))
+							i.scheduler.condition = cond
+							i.scheduler.condition.IncidentAge = age
+						} else {
+							i.scheduler.condition = nil
+						}
```
Comment (julianbrost) [inline internal/incident/incident.go:414]: Please explain what the value in `i.scheduler.condition` is supposed to express. I find it hard to follow what's going on here, `condition` stores an `EscalationFilter` which actually is more like a filter context (i.e. stores the context in which a filter is evaluated), but now it stores the time after which the filter is supposed to be reevaluated?

File: internal/incident/incidents.go [hunk 4]
```diff
@@ -20,6 +21,86 @@ var (
...
+
+			_, err = incident.evaluateRules(childCtx, tx, 0, types.Int{})
+			if err != nil {
+				return err
+			}
+
+			incident.evaluateEscalations()
+			if err = tx.Commit(); err != nil {
+				return err
+			}
```
[the remark this answers is not in the document]
Comment (yhabteab) [reply internal/incident/incidents.go:84]: No! This just loads all open incidents from the database and evaluates all their rules and escalation without waiting for external events of the incident object.

File: internal/incident/incident.go (hunk 3 above)
Comment (yhabteab) [reply internal/incident/incident.go:414]: `i.scheduler.condition` just stores the current escalation filter condition from which the `retryAfter` is extracted and modifies it's `IncidentAge` threshold and never stores `the time after which the filter is supposed to be reevaluated?`. This is needed because when you construct the escalation filter each time evaluating the escalations like:
https://github.com/Icinga/icinga-notifications/blob/14f00144b0fa82d5e40e70d3dbd70bb4588ff017/internal/incident/incident.go#L348

The condition won't match even it's retried after the correct **extracted** retry after interval.

**Comment by julianbrost** [comment]
Has a merge conflict now because I've merged #76 recently.

**Comment by yhabteab** [comment]
Used filter: `incident.age >= 2m`

```bash
2023-09-20T11:41:07.272+0200    INFO    listener        Processing event
2023-09-20T11:41:07.274+0200    INFO    incident        Opened incident {"object": "icinga2"}
2023-09-20T11:41:07.277+0200    INFO    incident        Source 1 severity changed from ok to crit       {"object": "icinga2", "incident": "#2"}
2023-09-20T11:41:07.280+0200    INFO    incident        Incident severity changed from ok to crit       {"object": "icinga2", "incident": "#2"}
2023-09-20T11:41:07.283+0200    INFO    incident        Rule "default" matches  {"object": "icinga2", "incident": "#2"}
```

**2 Minutes later:**
```bash
2023-09-20T11:44:07.253+0200    INFO    incident        Rule "default" reached escalation "[S] Office Hours"    {"object": "icinga2", "incident": "#2"}
2023-09-20T11:44:07.264+0200    INFO    incident        Notify contact "Yonas Habteab" via "rocketchat" {"object": "icinga2", "incident": "#2"}
2023-09-20T11:44:07.266+0200    ERROR   incident        Could not find config for channel       {"object": "icinga2", "incident": "#2", "type": "rocketchat"}
```

File: internal/incident/incidents.go [hunk 5]
```diff
@@ -19,6 +20,89 @@ var (
...
+			db:               db,
+			runtimeConfig:    runtimeConfig,
+			StartedAt:        incidentRow.StartedAt.Time(),
+			incidentRowID:    incidentRow.ID,
+			configFile:       configFile,
+			SeverityBySource: map[int64]event.Severity{},
+			EscalationState:  map[escalationID]*EscalationState{},
+			Rules:            map[ruleID]struct{}{},
+			Recipients:       map[recipient.Key]*RecipientState{},
+		}
```
Comment (julianbrost) [inline internal/incident/incidents.go:48]: You're duplicating the struct initialization here and it contains lots of "set to default values" (that primarily is initializing maps), so I think a `NewIncident()` function would help, otherwise changing anything becomes annoying (and if another required field is added, there's a good chance it might be forgotten in one place, which wouldn't happen if it was an argument of that `NewIncident()` function).

File: internal/incident/incidents.go [hunk 6]
```diff
@@ -19,6 +20,89 @@ var (
 	currentIncidentsMu sync.Mutex
 )
 
+// LoadOpenIncidents loads all active (not yet closed) incidents from the database and restores all their states.
+// Returns error ony database failure.
+func LoadOpenIncidents(
```
Comment (julianbrost) [inline internal/incident/incidents.go:25]: But in general, this function looks pretty similar to `GetCurrent()` and duplicates it quite a bit.

Shouldn't it be possible to simplify this function to something like this pseudocode:
```
for id in `select object_id from incident where recovered_at is null` {
  obj = LoadObjFromDB(id)
  incident = GetCurrentIncident(obj)
  incident.RestoreTimers()
}
```

Yes, that queries the `incident` table twice, but I'd rather keep the code simpler for now, as there'll be some quite intensive schema changes soon and we can optimize something like this after that was done.

File: internal/incident/incident.go (hunk 3 above)
Comment (julianbrost) [reply internal/incident/incident.go]: Please try to explain what these values are supposed to represent, without just explaining how they're set/accessed. So for example for the `Rules` member of the struct, that explanation would be something like "a set of all rules that are active on this incident and influence how notifications are sent" (instead of something like "rules are added when they match on this incident").

**Review by julianbrost — CHANGES_REQUESTED** [review]

File: internal/object/db_types.go [hunk 7]
```diff
@@ -27,3 +30,55 @@ type ObjectRow struct {
...
+	}
+
+	obj := &Object{db: db, ID: id, Tags: tags, Metadata: metadata}
+
+	cacheMu.Lock()
+	defer cacheMu.Unlock()
+
+	cache[id.String()] = obj
+
+	return obj, nil
```
[the remark this answers is not in the document]
Comment (yhabteab) [reply internal/object/db_types.go:83]: We don't need such checks here, I don't even think that mutex lock is required, since this method is ever called when starting the daemon and before the listener becomes ready. And as discussed yesterday, an object can only have a single active incident at a time, i.e this method will never load the same object multiple times from the database.

File: internal/incident/incident.go (hunk 3 above)
Comment (julianbrost) [reply internal/incident/incident.go]: After some more looking at this, I think you're trying to store the `EscalationFilter` context for the time in the future when the escalation conditions are due to be evaluated the next time. However, and that's probably what has irritated me, you don't seem to consider that other external events can happen in that time. So if you have an escalation for `incident_age>=2h` and after 1h, the incident goes from warning to crit, that has to affect the context that's used for evaluating the condition, however you seem to always use the already stored one.

File: internal/incident/incidents.go [hunk 8]
```diff
@@ -31,17 +94,9 @@ func GetCurrent(
...
-			runtimeConfig:    runtimeConfig,
-			configFile:       configFile,
-			Recipients:       map[recipient.Key]*RecipientState{},
-			EscalationState:  map[escalationID]*EscalationState{},
-			SeverityBySource: map[int64]event.Severity{},
-			Rules:            map[ruleID]struct{}{},
-		}
+		incident := NewIncident(db, runtimeConfig, configFile)
+		incident.Object = obj
+		incident.logger = logger.With(zap.String("object", obj.DisplayName()))
```
Comment (julianbrost) [inline internal/incident/incidents.go:99]: `object` and `logger` should be arguments of `NewIncident()` as well. After all, if they aren't set, that object won't function properly.

File: internal/incident/incident.go [hunk 9]
```diff
@@ -575,6 +640,115 @@ func (i *Incident) RestoreEscalationStateRules(states []*EscalationState) {
...
+		return
+	}
+
+	ctx := context.TODO()
+	if i.RestoreRecipients(ctx) != nil {
+		// The error is already logged within `i.RestoreRecipients()`
+		return
+	}
+
+	i.evaluateEscalations(i.StartedAt.Add(i.timer.incidentAgeFilter))
```
Comment (julianbrost) [inline internal/incident/incident.go:663]: Make that function return whether any new escalations were triggered. If there were none, you can return here and not send any notifications.

File: internal/incident/incident.go [hunk 10]
```diff
@@ -575,6 +640,115 @@ func (i *Incident) RestoreEscalationStateRules(states []*EscalationState) {
...
+		i.logger.Errorw("Failed to start time based incident escalations database transaction", zap.Error(err))
+
+		return
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	if i.notifyContacts(ctx, tx, &event.Event{Time: time.Now(), Type: event.TypeInternal}, types.Int{}) != nil {
+		// The error is already logged within `i.notifyContacts()`
+		return
+	}
```
Comment (julianbrost) [inline internal/incident/incident.go:676]: Change the event type to `escalation`. Then it also somewhat makes sense to send it to all contacts: for the new ones as a "hey, this incident was escalated to you" and for the old ones "hey, this incident was escalated to more people".

**Review by julianbrost — CHANGES_REQUESTED** [review]

**Comment by julianbrost** [comment]
For completeness also as a comment here: I've pushed a [branch](https://github.com/Icinga/icinga-notifications/tree/jbrost/time-based-escalation), that for one, rebases it onto the current `main` branch and also contains some draft commits that simplify the timer logic (there may be a timer for a reevaluation in the future, any evaluation of the escalations (including from new external events) just always cancel the timer and schedule a new one as needed).

**Comment by julianbrost** [comment]
I've just pushed an updated version with a bit cleaned up commits and rebased onto the current `main` to the aforementioned branch.

@yhabteab Can you please have a look if you can continue from here on again and make use of that branch for this PR?

Main open points should be at the moment:
- Who get's notified. I think you said that the functions you added in #99 will be helpful here to only contacts/channels that were newly escalated to.
- `evaluateEscalations()` needs a running database transaction now, but for the reevaluation, in case nothing new matches, we shouldn't need a transaction, because in that case, the transaction should be ended without executing a single query. So maybe we want to split that function, see also the inline TODO comment on my branch.

**Comment by yhabteab** [comment]
> evaluateEscalations() needs a running database transaction now, but for the reevaluation, in case nothing new matches, we shouldn't need a transaction, because in that case, the transaction should be ended without executing a single query. 

Would it be that expensive if we just keep it the way it is? We would have to sync the dummy event beforehand anyway, which seems not to be the case at the moment (in your branch). At worst, that tx would perform a single insert query and rollback it afterwards.

**Comment by julianbrost** [comment]
> Would it be that expensive if we just keep it the way it is?

I mean it's not prohibitively expensive, but it's also obvious room for improvement when we have all the information we need before starting the transaction.

> We would have to sync the dummy event beforehand anyway,

Do we have to? My idea idea would be to be to split `evaluateEscalations()` into two functions: one that doesn't need the database and simply evaluates the filters and returns which evaluations are to be triggered and another one to actually trigger escalations (without considering conditions, force trigger so to say). A bit similar to the recent rework in #99.

> which seems not to be the case at the moment (in your branch). 

Quite possible, I only looked at `type = 'notified'` events, those have the correct `event_id` set, others seem to have it missing.

**Comment by yhabteab** [comment]
I have reset this PR to your branch and based on that I did also implemented the missing parts you mentioned. From my tests everything seems to work, I just don't like what https://github.com/Icinga/icinga-notifications/pull/68/commits/60ba92edd6bb4305353ea0c32ee3e93ce7c761f5 is doing at the moment. These two commits have already cleaned it all up https://github.com/Icinga/icinga-notifications/pull/41/commits/57cd57981d196ac99dfde08b3260563ea2992eea, https://github.com/Icinga/icinga-notifications/pull/41/commits/76837ad789893b35f9a871e2b753df1bcbb78b3c, but you need to decide if you want to review  that PR or just merge this one and rework it once again in the other PR.

**Comment by julianbrost** [comment]
I don't want #41 to become a dependency for this PR. So if those two commits would work here on their own, we can consider it, otherwise this PR should probably better stay as it is.

File: internal/incident/incident.go [hunk 11]
```diff
@@ -320,61 +422,119 @@ func (i *Incident) evaluateEscalations(ctx context.Context, tx *sqlx.Tx, ev *eve
...
+
+						if !matched {
+							// Evaluate whether the escalation filter evaluates to true 1 nanosecond
+							// after the configured incident age condition.
+							return escalation.Condition.Eval(
+								&rule.EscalationFilter{IncidentAge: cond.IncidentAge + time.Nanosecond, IncidentSeverity: cond.IncidentSeverity},
+							)
+						}
+
+						return matched, nil
```
Comment (julianbrost) [inline internal/incident/incident.go:456]: I know that I suggested evaluating the rule at incident_age±1ns to capture strange uses of the `=` and `>` operator there, but now it feels like we're just making the code here more complex when we should probably tell users, that they should just go with `incident_age>=x` and maybe some others like `incident_age<x`. But for sure a restricted set of operators, because for `incident_age>=1h` it's obvious what should happen. Do we really want to make sure that `incident_age=1h` and `incident_age>1h` behave the same, or just offer `>=` in the web module for `incident_age`?

File: internal/incident/incident.go [hunk 12]
```diff
@@ -298,13 +387,26 @@ func (i *Incident) evaluateRules(ctx context.Context, tx *sqlx.Tx, eventID int64
...
 
+	// Escalations are reevaluated now, reset any existing timer, if there might be future time-based escalations,
+	// this function will start a new timer.
+	if i.timer != nil {
+		i.logger.Info("Stopping reevaluate timer due to escalation evaluation")
+		i.timer.Stop()
+		i.timer = nil
+	}
+
+	cond := &rule.EscalationFilter{IncidentAge: eventTime.Sub(i.StartedAt), IncidentSeverity: i.Severity}
```
Comment (julianbrost) [inline internal/incident/incident.go:405]: `cond` makes no sense for me here as a variable name. Should probably be something like `filterVars` (values for the variables that can be used in the escalation condition expression) or `filterContext` (context the escalation condition is evaluated in).

(I'm not a fan of that type being named `EscalationFilter` as it's not really a filter, but I don't want to start renaming it in this PR.)

File: internal/filter/parser_test.go [hunk 13]
```diff
@@ -62,101 +62,98 @@ func TestFilter(t *testing.T) {
...
 
 		rule, err = Parse("foo=bar|bar=foo")
 		assert.Nil(t, err, "There should be no errors but got: %s", err)
-		assert.IsType(t, &Any{}, rule)
+		assert.IsType(t, &Chain{}, rule)
 
 		rule, err = Parse("!(foo=bar|bar=foo)")
 		assert.Nil(t, err, "There should be no errors but got: %s", err)
-		assert.IsType(t, &None{}, rule)
+		assert.IsType(t, &Chain{}, rule)
```
Comment (julianbrost) [inline internal/filter/parser_test.go:113]: These should now also check `rule.op`.

File: internal/filter/parser_test.go [hunk 14]
```diff
@@ -62,101 +62,98 @@ func TestFilter(t *testing.T) {
...
 
 		rule, err = Parse("!foo")
 		assert.Nil(t, err, "There should be no errors but got: %s", err)
-
-		assert.Equal(t, &None{rules: []Filter{NewExists("foo")}}, rule)
+		assert.IsType(t, &Chain{}, rule)
```
Comment (julianbrost) [inline internal/filter/parser_test.go:117]: Why no more `assert.Equal()` here?

File: internal/filter/parser_test.go [hunk 15]
```diff
@@ -62,101 +62,98 @@ func TestFilter(t *testing.T) {
...
-			&None{rules: []Filter{
-				&Equal{column: "foo", value: "bar"},
-				&Equal{column: "bar", value: "foo"},
-			}},
-			&Any{rules: []Filter{
-				&Equal{column: "foo", value: "bar"},
-				&Equal{column: "bar", value: "foo"},
-			}},
-		}}
-		assert.Equal(t, expected, rule)
```
Comment (julianbrost) [inline internal/filter/parser_test.go:129]: Why was this test removed altogether?

File: internal/filter/types.go [hunk 16]
```diff
@@ -1,67 +1,106 @@
...
 
-	return true, nil
-}
+const (
+	// NONE represents a filter chain type that matches when none of its ruleset matches.
+	NONE LogicalOp = "!"
+	// ALL represents a filter chain type that matches when all of its ruleset matches.
+	ALL LogicalOp = "&"
+	// ANY represents a filter chain type that matches when at least one of its ruleset matches.
+	ANY LogicalOp = "|"
```
Comment (julianbrost) [inline internal/filter/types.go:16]: Why are these ALL CAPS in contrast to the `CompOperator` constants?

File: internal/filter/types.go
Comment (julianbrost) [inline internal/filter/types.go:1]: `NewCondition` seems unused now and also seems incomplete now and should therefore better be removed (or updated and used everywhere a `Condition` is created).

File: internal/filter/types.go [hunk 17]
```diff
@@ -73,116 +112,108 @@ func NewCondition(column string, value string) *Condition {
...
+	case UnEqual:
+		match, err = filterable.EvalEqual(c.column, c.value)
+		if err != nil {
+			return false, err
+		}
 
-func NewExists(column string) *Exists {
-	return &Exists{column: column}
-}
+		match = filterable.EvalExists(c.column) && !match
```
Comment (julianbrost) [inline internal/filter/types.go:132]: I find it surprising that `EvalEqual` is done before `EvalExists`. But at the very least, it should be `!match && EvalExists()`, no need to run that function if it can't influence the result anyways.

**Review by julianbrost — CHANGES_REQUESTED** [review]

File: internal/filter/parser_test.go (hunk 13 above)
Comment (yhabteab) [reply internal/filter/parser_test.go:113]: Just ignore theses! There's an extended test case in the other PR.

File: internal/rule/condition.go [hunk 18]
```diff
@@ -3,82 +3,111 @@ package rule
...
 }
 
-func (c *EscalationFilter) EvalEqual(key string, value string) (bool, error) {
+// ExtractRetryAfter extracts a time.Duration from the given filter conditions.
+// The retry after duration is extracted from the specified filter conditions that evaluate the "incident_age"
+// filter column. This function returns rule.RetryNever when all the incident age filter values are less than
+// the actual incident age or none of the filter conditions evaluates the "incident_age" column. Otherwise,
+// a time.Duration < rule.RetryNever and > the current incident age at which this escalation condition should
+// be re-evaluated is returned.
+func (e *EscalationFilter) ExtractRetryAfter(conditions []filter.Conditioner, incidentAge time.Duration) time.Duration {
```
Comment (julianbrost) [inline internal/rule/condition.go:25]: This should either not be implemented on `*EscalationFilter` or not have the `incidentAge` parameter.

**Review by julianbrost — CHANGES_REQUESTED** [review]

File: internal/filter/contracts.go [hunk 19]
```diff
@@ -12,4 +12,12 @@ type Filterable interface {
...
 	Eval(filterable Filterable) (bool, error)
+	ExtractConditions() []Conditioner
+}
+
+// Conditioner provides a way to access a filter condition column and value.
+// It is implemented by the Condition and Exists filter types.
+type Conditioner interface {
+	Column() string
+	Value() string
 }
```
[the remark this answers is not in the document]
Comment (yhabteab) [reply internal/filter/contracts.go:23]: > That interface feels like it's not necessary at all. After all, it's only used by `ExtractConditions()` which in turn is used to get all values `incident_age` is compared with.

How do you think the fields from the `EscalationFilter#ExtractRetryAfter()` method should be accessed without making them exportable then? Yes, of course the filter `Exists` does not have to implement this interface, but I don't see any disadvantages in doing so either.

[the remark this answers is not in the document]
Comment (julianbrost) [reply internal/filter/contracts.go:23]: > How do you think the fields from the `EscalationFilter#ExtractRetryAfter()` method should be accessed without making them exportable then?

Would there be anything from with just exporting them?

> Yes, of course the filter `Exists` does not have to implement this interface, but I don't see any disadvantages in doing so either.

There is no need for it and the the current implementation claims that the column/variable `incident_age` is compared with the value `incident_age` which just isn't the case.

[the remark this answers is not in the document]
Comment (yhabteab) [reply internal/filter/contracts.go]: Just removed the interface but still didn't make the fields exportable.

**Review by julianbrost — CHANGES_REQUESTED** [review]

File: internal/rule/condition.go [hunk 20]
```diff
@@ -3,82 +3,112 @@ package rule
...
 
-func (c *EscalationFilter) EvalEqual(key string, value string) (bool, error) {
+// ExtractRetryAfter extracts a time.Duration from the given filter conditions relative to the actual incident age.
+//
+// The retry after duration is extracted from the specified escalation condition that evaluate the "incident_age"
+// filter column. This function returns rule.RetryNever when all the incident age filter values are less than
+// the actual incident age or none of the filter conditions evaluates the "incident_age" column. Otherwise,
+// a time.Duration < rule.RetryNever and > the current incident age at which this escalation condition should
+// be re-evaluated is returned.
+func (e *EscalationFilter) ExtractRetryAfter(escalationCond filter.Filter) time.Duration {
```
Comment (julianbrost) [inline internal/rule/condition.go:26]: My suggestion for the overall function documentation (including a renaming of the function itself) would be the following:

```suggestion
// ReevaluateAfter returns the duration after which escalationCond should be reevaluated the 
// next time on the incident represented by e.
//
// escalationCond must correspond to an escalation that did not trigger on the incident
// represented by e before. If nothing in the incident changes apart from time passing by,
// the escalation is guaranteed to not trigger within the returned duration. After that
// duration, the escalation should be reevaluated and it may or may not trigger. If anything
// else changes, for example due to an external event, the escalation must be reevaluated as
// well.
func (e *EscalationFilter) ReevaluateAfter(escalationCond filter.Filter) time.Duration {
```

Parts of the current comment could be moved to inline comments within the implementation to explain it if you deem that necessary. But my impression is that it just rephrases the code in human language without giving much extra explanation.

File: internal/rule/condition.go [hunk 21]
```diff
@@ -3,82 +3,114 @@ package rule
...
+	for _, condition := range escalationCond.ExtractConditions() {
+		if condition.Column() == "incident_age" {
+			incidentAgeFilterValue, err := time.ParseDuration(condition.Value())
+			if err != nil {
+				continue
+			}
+
+			if incidentAgeFilterValue > e.IncidentAge {
+				retryAfter = min(retryAfter, incidentAgeFilterValue)
+			}
```
Comment (julianbrost) [inline internal/rule/condition.go:39]: D'oh, in my head, we already changed the return value to be relative to `e` (~now) instead of the incident start. Thus, I've written the comment with that in mind, but now it's actually inconsistent with the implementation.

Anyways, I'd change the implementation to match the comment, because I think this makes sense that way around. It can also be written a bit more compact without becoming harder to understand IMHO, maybe even a little improvement in that regard with no more `continue`.

```suggestion
			v, err := time.ParseDuration(condition.Value())
			if err == nil && v > e.IncidentAge {
				// The incident age is compared with a value in the future. Once that age is
				// reached, the escalation could trigger, so consider that time for reevaluation.
				retryAfter = min(retryAfter, v - e.IncidentAge)
			}
```

Of course, also requires a change where it's called to use the return value in the appropriate way.

**Review by julianbrost — CHANGES_REQUESTED** [review]

File: internal/rule/condition.go (hunk 21 above)
Comment (yhabteab) [reply internal/rule/condition.go:39]: I somehow can't quite grasp why you keep bringing up how complex this is. It's only 2-3 lines of code that you don't need any comments to be able to understand it. But hey, that shouldn't matter, even if I didn't understand your comments from yesterday enough to understand what the function does compared to the previous ones. The comments just describe what might or might not happen when using this function somewhere in incident package.

Comment (julianbrost) [reply internal/rule/condition.go:39]: The main change request here is to do this, i.e. change the reference point for the return value to match the updated documentation:
```suggestion
			incidentAgeFilterValue, err := time.ParseDuration(condition.Value())
			if err != nil {
				continue
			}

			if incidentAgeFilterValue > e.IncidentAge {
				retryAfter = min(retryAfter, incidentAgeFilterValue - e.IncidentAge)
			}
```

But then I thought this can be simplified a bit and proposed a larger change.

Regarding the function documentation and inline comments: I think that those should give context to the code, they should say why and what for things are done. The function we're talking here has a very specific purpose: if I have an incident and an escalation that hasn't triggered yet, tell me for how long is it safe to ignore that escalation, because the result of evaluating the condition won't change until then. From the view of someone calling the function, that should be the relevant information, not what values are compared with the incident age in which way. I was trying to push you incrementally towards a direction where I think that is easier to understand from the interface and documentation of the function. I thought just requesting a big "just do it completely differently" change would have been frustrating for you, but my approach is already taking more iterations and time than I had expected, so it wasn't very successful either.

File: internal/incident/incident.go [hunk 22]
```diff
@@ -320,61 +426,89 @@ func (i *Incident) evaluateEscalations(ctx context.Context, tx *sqlx.Tx, ev *eve
...
+		i.logger.Infow("Scheduling escalation reevaluation", zap.Duration("after", retryAfter), zap.Time("at", nextEvalAt))
+		i.timer = time.AfterFunc(retryAfter, func() {
+			i.logger.Info("Reevaluating escalations")
 
-						return errors.New("failed to upsert escalation state")
-					}
+			i.RetriggerEscalations(&event.Event{
+				Type:    event.TypeInternal,
+				Time:    nextEvalAt,
+				Message: fmt.Sprintf("Incident reached age %v", retryAfter),
```
Comment (julianbrost) [inline internal/incident/incident.go:465]: This message is now incorrect with the changed value: If you have escalations `incident_age>=1m` and `incident_age>=2m`, the second escalation will also use "age 1m" because that's the difference to the last evaluation.

**Review by julianbrost — CHANGES_REQUESTED** [review]
One message in the event is incorrect. Apart from that, this PR should be good to go.

**Review by julianbrost — APPROVED** [review]
